### PR TITLE
Refactor of current style PR so spackbot can fix style

### PR DESCRIPTION
**Pull request by vsoch**
This PR will replace #11 . Importantly, I haven't been able to test this fully yet because I need ssh credentials for spackbot. So remaining work to do:

1. get credentials
2. test functionality
3. take screenshots and update docs
4. update developer docs for creating and using credentials

I am likely to change the workflow a bit when there are credentials, although I can't say for sure until I test.

ping @tgamblin @adamjstewart !

Update 07/19: tested and ready for review.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

### CODE REVIEW

File: spackbot/handlers/style.py
Comment (tgamblin) [inline spackbot/handlers/style.py:100]: This runs code from the PR, which is a security hole.  The PR author can change the `spack style` code in their PR to run anything they want as `spackbot`.

I think one way to do this is:

1. Clone `develop` first
2. Add the `develop` clone to `PATH`, then `from sh import spack`
3. Make a copy of the cloned repo
4. Add `clone_url` (which should probably be called `fork_url` or something since we're not cloning it now) as a remote in the copy
5. Check out `branch` in the copy (the one that is *not* in `PATH`) to avoid a second full clone (which is slow)
6. Run `spack style --root .` in the clone

I'm adding a `--root` option to `spack style` so that we can do this.  It will make the following changes:

* Run in the clone's prefix
* Be sure to use the invoked `spack` instance's configs for the various tools (i.e., don't allow `flake8`, `isort`, `black`, `mypy`, etc. to use configs from the clone (`flake8`'s config in particular can specify additional python code files to run)

We should probably also have spackbot decline to format PRs that modify style configs or the `spack style` command, as they won't work if we're only running `spack style` from `develop` -- users making changes to those would need to run `spack style` themselves.

Having written all that, the other option that I kind of like a lot better would be to have `spackbot` run `spack style` in a container as an unprivileged user.  Do you know how to do that @vsoch?  We could bindmount the clone and run its `spack style` in an environment that does not have access to the `spackbot`'s credentials. If we could do that we could avoid some of this complexity.  Basically I do not want the code from the PR to be able to get at private keys, tokens, `id_spackbot`, etc.

@adamjstewart @becker33 @alalazo thoughts on this?

File: .github/workflows/main.yml
Comment (tgamblin) [inline .github/workflows/main.yml:26]: Can you please add `flake8` here, with config that matches the black config? See https://black.readthedocs.io/en/stable/guides/using_black_with_other_tools.html?highlight=flake8#flake8

`flake8` will catch the stuff that `black` can't fix -- e.g. there are some lines in this PR that wouldn't pass `flake8` but `black` currently doesn't break up long strings.

**Review by tgamblin — CHANGES_REQUESTED** [review]

File: docs/docs/developer-guide/docker-compose.md
Comment (vsoch) [reply docs/docs/developer-guide/docker-compose.md:41]: Sure I'll remove.

File: spackbot/handlers/style.py
Comment (vsoch) [reply spackbot/handlers/style.py:71]: I think either will still work for the commit, at least I've had both work for me. Are you sure we need to add this extra endpoint?

File: spackbot/handlers/style.py
Comment (vsoch) [reply spackbot/handlers/style.py:83]: I thought we talked about this, and we wanted the committer to be the person making the PR?

File: .github/workflows/main.yml
Comment (vsoch) [reply .github/workflows/main.yml:26]: I think this is out of scope for this PR, probably adding this file is too (that's my fault).

File: spackbot/handlers/style.py
Comment (tgamblin) [reply spackbot/handlers/style.py:83]: I thought we said the author. So question is really whether author and committer both or just author.

File: spackbot/handlers/style.py
Comment (tgamblin) [reply spackbot/handlers/style.py:71]: If both work for you then I think we could unconditionally use the id -- your account was created before July 18, 2017.

If we omit the id, it can end up that we do the commit on behalf of the wrong user (for users where there was some previous incarnation of themselves).

I think that we need to get the user's id somehow regardless.  Curious if you can get that from the `pr` dictionary.

Comment (vsoch) [reply spackbot/handlers/style.py]: Ok this is added.

**Comment by vsoch** [comment]
PR is updated so that:
 - we don't keep adding to the path but target an executable directly (in style and reviewers)
 - spack is cloned from develop, we run style from there (and not from the PR). 
 - this means that we don't need to check if the version of spack is too old to have style (it always will)
 - the user email is looked up to derive the correct version of the email based on the creation date.

And now I'm having dinner because I'm getting hangry!!! :smiling_imp:

**Comment by vsoch** [comment]
@adamjstewart @alalazo your feedback would be useful before I refactor this again, @tgamblin has suggested that the bot should:

- clone develop and copy pyproject.toml off somewhere
- check out the branch from the fork
- do what spack style does to get the list of changed files (git diff develop… )
- run isort and black on the changed files with --config /path/to/develop/pyproject.toml
- commit stuff and push

instead of the current approach. What are your thoughts? I want to make sure we're all on the same page.

File: .github/workflows/main.yml
Comment (tgamblin) [reply .github/workflows/main.yml:26]: ok sure -- this can be separate

**Comment by adamjstewart** [comment]
I think we're all in agreement as to what spack-bot should do, the question is more how to do it. I think @tgamblin is right about the security vulnerability, but I know next to nothing about security, so I'll defer to his judgement.

**Comment by vsoch** [comment]
What are we getting out of manually writing again the commands in spack style vs. just using spack style? This is really arduous. And if there are ever any changes in spack style we would have to update the bot to stay in sync, and what happens if the two return a different result? I'm not liking this approach.

**Comment by adamjstewart** [comment]
Is there any way we can prevent a security vulnerability other than doing everything manually? This is also annoying for maintainer pinging because it thinks all new Python packages have no maintainer (even though I'm listed as a maintainer in the base class).

**Comment by vsoch** [comment]
I guess I don't understand why running spack style from a develop branch is any different from doing the same for spack maintainers.

**Comment by adamjstewart** [comment]
If you're on the develop branch, running `spack style` won't do anything since there are no changed files. You have to be on the PR branch, where users can modify the code to do whatever they want. Could we add an option to `spack style` that changes the root directory of Spack? That way we are running `spack style` with Spack's develop branch and running it on a different clone in a different directory on the PR branch.

**Comment by vsoch** [comment]
@adamjstewart we are doing two clones - one of the develop (and we add this spack to the path) and then the second is where we run the command.

**Comment by vsoch** [comment]
> Could we add an option to spack style that changes the root directory of Spack? That way we are running spack style with Spack's develop branch and running it on a different clone in a different directory on the PR branch.

Yeah I think @tgamblin was saying he was doing this but then maybe he ran into something where he decided it was a bad idea because the plan changed to this current list, which feels too complicated to me.

**Comment by vsoch** [comment]
ug, again? What was merged into main (and how) to totally screw up all my PRs (note that it didn't work to fix it locally) :sob:

**Comment by tgamblin** [comment]
Reopening this and closing #29 to keep the discussion in one place.  The issue was that the branch can't be rebased b/c it has merge commits:

Screen Shot 2021-07-25 at 1 08 37 AM

When that happens you just want to squash the commits, but we didn't have squashing enabled on the repo for some reason -- guess I didn't click that checkbox. Now it's squash-mergeable:

Screen Shot 2021-07-25 at 1 07 52 AM

Choosing rebase will still complain b/c there's no way to replay the commits linearly. I don't usually enable merge commits for PRs b/c I like linear history, but that would've worked as well.  As you can see from the actual history though it would be nasty.

**Comment by tgamblin** [comment]
Oh also it looks like #29 is just a squashed version of this, so we could force-push that here.  I think @vsoch made it by hand but the way I typically do that if I just want to get rid of ugly history is:

```console
git reset main   # base new commit on this branch but stay on add/style-again
git commit       # commit all the changes
```

Then force push.

**Comment by tgamblin** [comment]
Ok back to the discussion.  We need to decide what to do.

## Security stuff

Just to clarify, the security concern is basically:
* We cannot run code from a PR in a privileged environment.

Running checks from within Spackbot is privileged because:
* It has many secrets we don't want PR authors to exfiltrate. 
* It also has write permissions on the main Spack repo (which it needs in order to push commits to PRs)

Why do we care? The code from the PR can be anything. PR authors could modify `spack style` to push anything to Spack's `develop` branch. We can't do something that would make that possible.

The security posture we're taking with Spack, as an open source project, is that we trust our maintainers (people with write access to Spack and with PR merge permissions).  If they merge something, we can run it privileged.  They can make mistakes, but that's a risk that everyone who uses Spack (or any OSS project) signs up for.  We've decided the cost of not using OSS outweighs that risk.

PRs are not approved by maintainers, but `develop` is.  So *we can run code from `develop`* in Spackbot, but not code from PRs that hasn't been vetted.

## Options
These are what I see as the options:

### 1. Run `spack style` unprivileged

We already run `spack style` in GitHub actions -- we've actually run the style check by the time Spackbot might want to fix style for a PR.  We could just use the output of that.  That would look like:

1. Run `spack style --fix` in the action.
2. If fixes were needed, have the action run `git format-patch`, save the result as an artifact, and fail the check.
3. Spackbot notices and says "hey do you want to fix that?"
4. Make `@spackbot fix style` download the patch and apply it with `git am`, then push the result.

Advantages:
* Applying a patch in the bot is simple and secure -- the bot never runs insecure code.
* We can run the most current `spack style` because we run it in an action, where it can't do any harm.

Disadvantages:
* The style check has to pass before you can tell `spackbot` to fix style.
* This makes it very tightly coupled to Spack's CI and harder to debug (can't just debug on an arbitrary repo)
* Complicated in that it requires us to fetch artifacts from actions.


### 2. Run `spack style` from `develop`

In this case we'd need to:

1. Add a `--root` option to `spack style` so that it can be pointed to a different prefix.
2. Make `spack style` run all tools with the `--config` option so that we only use configs from `develop`.  This might seem like overkill, but we don't want the tools to take configuration from the PR.  Things like the `.flake8` config allow you to specify plugins, and we actually use that. `black`'s config seems very minimal and safe, but `isort` seems to allow plugins like `flake8`. I worry that this introduces a potential way for the PR to run code in `spackbot`.

With those two changes, I think `spackbot` can check out both `develop` and the PR branch, and run spack style from `develop` on the fork.

Advantages:
* No artifacts/saving patches
* Uses `spack style` mostly the way it was intended

Disadvantages:
* `spack style` is run from `develop` so the bot won't see changes to style config *until* they make it to `develop`.  People changing style will need to run `spack style` manually.
* Not as obviously secure as (1) -- if we someone fail to force the config to point to `develop` things could go wrong.

### 3. Run `isort` and `black` in the bot (reimplement `spack style`)

I thought about this one some more and the original motivation was to avoid running code from the PR. But with the `isort` config issue, it has the same issues as (2), so I don't think it's really worth the reimplementation.

### 4. Run `spack style` in a sandbox

You can sandbox things with docker -- we could have `spackbot` do that. But I googled around and couldn't find anything terribly simple for doing this. `sysbox` looks promising but I don't get the ins and outs and don't think we need to.


## Ok so what do we do?

Ok @alalazo @vsoch @adamjstewart: given all that I think either (1) or (2) is the best choice.  Any opinions?

**Comment by vsoch** [comment]
My preference is for 2 because:

1. we don't want to be running something different that possibly could deviate from what a `spack style --fix` would do (so this eliminates 1 and 3 for me)
2. I'm not sure how running black and isort without spack style is really that different, it just adds the issue above.
3. the container approach would be much harder because development environments (and spackbot) are already run in containers. (eliminates 4)

And that means we are left with 2, which I like because:
1. It mirrors exactly what the user would do locally. A change / PR to spack style itself is an edge case, and in this case, the developer can run it locally (the default right now)
2. we are already running spack maintainers (and if this was okay, spack style should be too, especially being run from spack develop)
3. the current PR is already setup to do this, we just need the `--root` and change to target configs.

**Comment by tgamblin** [comment]
Ok sounds good (I like 1 and 2), but just to make sure:

> we don't want to be running something different that possibly could deviate from what a spack style --fix would do (so this eliminates 1 and 3 for me)

(1) runs `spack style —fix` exactly like the user would; it just uses git to save the results  from the action and lets the bot apply them. So I don’t understand why you eliminate it for *this particular* reason. Does that change anything for you? Not saying it doesn’t have its own downsides.

**Comment by vsoch** [comment]
Nope, no change. I still think 1 is too complex, and although technically it should do the same thing, because it is not exactly the same thing (there are extra steps) I can’t eliminate the possibility that something could go wrong I just don’t know about.

**Comment by tgamblin** [comment]
Ok yay! I'm going to add the `--root` option to `spack style`.

**Comment by tgamblin** [comment]
Ok `--root` is done! https://github.com/spack/spack/pull/25085

It should be pretty straightforward to use `spack style` in the bot now!

**Comment by vsoch** [comment]
Will update soon!! (about to go into meeting)

**Comment by vsoch** [comment]
Root is added: https://github.com/spack/spack-bot/pull/22/commits/a1fcfb50f28c19011ede7ec228508bbc3b33f89f

**Comment by vsoch** [comment]
I hope that worked! :grimacing:

File: spackbot/handlers/style.py
Comment (tgamblin) [inline spackbot/handlers/style.py:105]: Instead of doing another deep clone and another fetch of `develop` here, make a local copy of the `spack-develop` you already cloned, then fetch the new branch (it's a lot faster because you already have the full history locally):

Instead of doing another deep clone here, make a local copy of `spack-develop` and just fetch the new branch (it's a lot faster because you already have the full history locally):

```suggestion
        # existing clone already has `develop` fetched and checked out
        shutil.copy("spack-develop", "spack")
        os.chdir("spack")
        git("config", "user.name", user)
        git("config", "user.email", email)

        # This will authenticate the push with the added ssh credentials
        git("remote", "rename", "origin", "upstream")
        
        # probably want to rename clone_url to fork_url or something
        # since we don't clone it anymore
        git("remote", "add", "origin", clone_url)

        # we're on upstream/develop. Fetch and check out just the PR branch
        git("fetch", "origin", f"{branch}:{branch}")
        git("checkout", branch)
```

Note that this probably seems more natural for the last two lines:

```python
        git("fetch", "origin")
        git("checkout", branch)
```

But I think that won't work in the case that the branch name exists *both* in `upstream` and `origin` -- it'll error ou and say it's ambiguous.  Doing the explicit fetch gets you just that branch and it creates the local branch so that the checkout is unambiguous.

File: spackbot/handlers/style.py
Comment (tgamblin) [inline spackbot/handlers/style.py:121]: Mostly minor but I would write this stuff using the error code exceptions in `sh`:

```suggestion
        try:
            git("commit", "-a", "-m", f"[spackbot] updating style on behalf of {user}")
        except sh.ErrorReturnCode_1 as e:
            # `commit` returns 1 if the branch is up to date
            if not is_up_to_date(e.stdout):
                raise  # something else went wrong
                
            return message + (
                "\nI wasn't able to make any further changes, "
                "but please see the message above for remaining "
                "issues you can fix locally!"
            )
                
        message += "\nI've updated the branch with isort fixes."
```

**Review by tgamblin — CHANGES_REQUESTED** [review]
Ok I went over this in detail -- the logic looks good to me, but I have some change requests.

File: spackbot/routes.py
Comment (vsoch) [reply spackbot/routes.py:103]: Sure, but I think this is so spread over the code that it should be a separate PR - it's out of scope of the style changes here.

File: spackbot/helpers.py
Comment (vsoch) [reply spackbot/helpers.py:58]: Agree that it's out of scope. I used requests, however, because using the github api handle didn't work for this simple request. It may have been that it didn't require authentication and the client always had it.

File: spackbot/helpers.py
Comment (vsoch) [reply spackbot/helpers.py:36]: yeah we can try that, agree another PR.

File: spackbot/handlers/style.py
Comment (vsoch) [reply spackbot/handlers/style.py:50]: I think we already discussed this is out of scope and can be tried with another PR - the requests is uesd for a reason (I don't think the async call with the authenticated client worked here) so I defaulted back to simple. KISS principle!

File: .flake8
Comment (vsoch) [reply .flake8:1]: I don't think we run flake8 anywhere.

File: spackbot/comments.py
Comment (vsoch) [reply spackbot/comments.py:67]: Out of scope, another PR.

File: spackbot/handlers/style.py
Comment (vsoch) [reply spackbot/handlers/style.py:59]: There are different actions taken depending on finding (or not finding) the contributor, and the actual function is just that one line with different variables too, so I think this is better as it currently is. There isn't enough commonality beyond that one lie, which is already a function.

File: spackbot/handlers/style.py
Comment (vsoch) [reply spackbot/handlers/style.py:79]: Sure thing.

File: spackbot/handlers/style.py
Comment (vsoch) [reply spackbot/handlers/style.py:85]: Yep

File: spackbot/handlers/style.py
Comment (vsoch) [reply spackbot/handlers/style.py:105]: I don't know why, but this doesn't work - it freezes on the clone and spackbot stops responding.

File: spackbot/handlers/style.py
Comment (vsoch) [reply spackbot/handlers/style.py:121]: This doesn't seem to work and I'm not liking the "raise" case - I think we always want spackbot to respond with something.

File: spackbot/handlers/style.py
Comment (tgamblin) [reply spackbot/handlers/style.py:50]: I’d like you to try to figure this one out. Importing another library to do something aiohttp should be able to do is not KISS IMO. We should figure out the deeper problem instead of doing a workaround like this. Not seeing why the authenticated request would fail because of authentication.

File: .flake8
Comment (tgamblin) [reply .flake8:1]: Right. But the setting are there so you can locally — it’s just not in CI

File: spackbot/handlers/style.py
Comment (vsoch) [reply spackbot/handlers/style.py:50]: yes but requests is used throughout the library in other handlers not relevant to this PR. If it needs to be fixed that's okay but it's out of scope for here.

Comment (vsoch) [reply spackbot/handlers/style.py:50]: > Not seeing why the authenticated request would fail because of authentication.

There are many cases in the wild where adding authentication actually makes it not work. Not sure if that is the issue here, but that's definitely a thing.

File: .flake8
Comment (vsoch) [reply .flake8:1]: Sure I'll add it back, if it's important to you.

Comment (tgamblin) [reply .flake8]: Yeah it would be kind of ironic for the style bot to have no style checks. We can add it to CI in another PR.

Comment (vsoch) [reply .flake8]: He's a rebel.

**Comment by vsoch** [comment]
I opened issues for the future changes requested that are out of scope for this PR.

File: spackbot/handlers/style.py
Comment (tgamblin) [reply spackbot/handlers/style.py:121]: Yeah could omit the `raise` -- I modified it above.  I'm fine with ignoring this and using your helper though.

File: spackbot/handlers/style.py
Comment (tgamblin) [reply spackbot/handlers/style.py:105]: What part doesn't work?  Cloning via `ssh`?  Confused because there is no clone in my modified version -- the initial clone is the same.

Comment (tgamblin) [reply spackbot/handlers/style.py:105]: Did you mean that this clone doesn't work? https://github.com/spack/spack-bot/pull/22#discussion_r678384010

Comment (vsoch) [reply spackbot/handlers/style.py:105]: I don't know what the issue was, but I'd trigger the bot in the PR and I would see the clone command in the logs and then it would just freeze like that. The ssh change in another comment worked fine, but this particular chunk did not.

Comment (tgamblin) [reply spackbot/handlers/style.py:105]: Can you debug it and see where it is actually hanging?

Comment (vsoch) [reply spackbot/handlers/style.py:105]: Hung here for like 5 mins

image

And previously I had killed it but I let it run this time, and I arrived at this (and note I had to change shutil.copy to shutil.copytree because copy is for a file, not directory, and now we have some recursive thing going on:

image

By the way - this is just the top of the screen - the output is like 10x longer and then it finally issues a 500 internal server error. So probably shutil copytree of the repository is a bad idea? lol

Comment (vsoch) [reply spackbot/handlers/style.py:105]: see "too many levels of symbolic links." We would need to do some kind of custom copy strategy which sounds overly complicated imho.

My thinking is that it works well as is currently implemented, this seems a bit buggy, so we should keep / try out as is until there is evidence that it's not good enough or needs to be changed.

Comment (tgamblin) [reply spackbot/handlers/style.py:105]: The bug is really simple.  Change this:

```python
shutil.copy("spack-develop", "spack")
```

to this:

```python
shutil.copy("spack-develop", "spack", follow_symlinks=False)
```

Comment (tgamblin) [reply spackbot/handlers/style.py:105]: There is a symlink in `lib/spack/docs` that points up to the top-level spack directory, so it's just recursive.  I don't know why `shutil` follows symlinks by default -- that seems like a bad default... but we've run into that before.  Sigh.

Comment (vsoch) [reply spackbot/handlers/style.py:105]: copy is for files, I think we want:

```python
# existing clone already has `develop` fetched and checked out
shutil.copytree("spack-develop", "spack", symlinks=False)
```

Comment (tgamblin) [reply spackbot/handlers/style.py:105]: Reworked fix to use `copytree` as you pointed out.

Comment (vsoch) [reply spackbot/handlers/style.py:105]: It still takes a lot longer than the other way, and now I'm seeing a new error that I don't know what is.
```bash
ot_1  | INFO:spackbot:Received event <gidgethub.sansio.Event object at 0x7ff4f40cf650>
spackbot_1  | INFO:sh.command:<Command '/usr/bin/git clone --depth 1 git@github.com:spack/spack spack-develop', pid 17>: process started
smee_1      | Event { type: 'error', message: undefined }
smee_1      | Connected https://smee.io/xxxxxxxxxx
```
I still think we should keep the current way.

Comment (tgamblin) [reply spackbot/handlers/style.py]: Ok anyway we resolved this offline and got rid of shutil (python copying is slow) and eliminated one remote clone.  yay!

Comment (vsoch) [reply spackbot/handlers/style.py]: yes yay! TLDR: shutil is just really terrible.

**Review by tgamblin — APPROVED** [review]

**Comment by tgamblin** [comment]
Ok @vsoch -- LGTM!  Thanks for all the iteration on this!

**Pull request merged**
